src/ucs/datastruct/ptr_map.inl

```diff
@@ -79,36 +154,57 @@ static inline void ucs_ptr_map_destroy(ucs_ptr_map_t *map)
  * @return UCS_OK on success, UCS_ERR_NO_PROGRESS if this key is direct and
  *         therefore no action was performed, otherwise error code as defined by
  *         @ref ucs_status_t.
- * @note @ptr must be aligned on @ref UCS_PTR_MAP_KEY_MIN_ALIGN.
+ * @note @a ptr must be aligned on @ref UCS_PTR_MAP_KEY_MIN_ALIGN.
  */
 static UCS_F_ALWAYS_INLINE ucs_status_t
 ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr, int indirect,
                 ucs_ptr_map_key_t *key)
 {
-    khiter_t iter;
-    int ret;
-
-    if (ucs_likely(!indirect)) {
-        *key = (uintptr_t)ptr;
-        ucs_assert(!(*key & UCS_PTR_MAP_KEY_MIN_ALIGN));
-        ucs_assert(*key != 0);
+    if (ucs_likely(ucs_ptr_check_direct_put(ptr, indirect, key))) {
         return UCS_ERR_NO_PROGRESS;
     }
 
-    *key = (map->next_id += UCS_PTR_MAP_KEY_MIN_ALIGN) |
-           UCS_PTR_MAP_KEY_INDIRECT_FLAG;
+    *key = ucs_ptr_map_create_key(map);
+    return ucs_ptr_hash_put(&map->hash, *key, ptr);
+}
 
-    iter = kh_put(ucs_ptr_map_impl, &map->hash, *key, &ret);
-    if (ucs_unlikely(ret == UCS_KH_PUT_FAILED)) {
-        return UCS_ERR_NO_MEMORY;
-    } else if (ucs_unlikely(ret == UCS_KH_PUT_KEY_PRESENT)) {
-        return UCS_ERR_ALREADY_EXISTS;
+
+/**
+ * Put a pointer into the map. Thread safe version of @ref ucs_ptr_map_put.
+ *
+ * @param [in]  map       Container.
+ * @param [in]  ptr       Object pointer.
+ * @param [in]  indirect  If nonzero, the pointer @a ptr is stored in the
+ *                        internal hash table, associated with a unique indirect
+ *                        key which is returned from this function. Otherwise,
+ *                        the returned value is the integer representation of
+ *                        the pointer @ptr.
+ * @param [out] key       Key to object pointer @ptr if operation completed
+ *                        successfully otherwise value is undefined.
+ * @return UCS_OK on success, UCS_ERR_NO_PROGRESS if this key is direct and
+ *         therefore no action was performed, otherwise error code as defined by
+ *         @ref ucs_status_t.
+ * @note @a ptr must be aligned on @ref UCS_PTR_MAP_KEY_MIN_ALIGN.
```

Comment:
maybe it's worth mentioning that it may not be safe to use it together with `ucs_ptr_map_put`